src/services/obs/backend.rs

```diff
@@ -0,0 +1,310 @@
+};
+use crate::ops::BytesRange;
+use crate::services::util;
+use crate::Scheme;
+use crate::{
+    http_util::HttpClient,
+    ops::{OpCreate, OpDelete, OpList, OpRead, OpStat, OpWrite},
+    Accessor, AccessorMetadata, BytesReader, BytesWriter, DirStreamer, ObjectMetadata,
+};
+
+/// Builder for Huaweicloud OBS services
+#[derive(Default, Clone)]
+pub struct Builder {
+    root: Option<String>,
+    endpoint: Option<String>,
+    access_key: Option<String>,
+    secret_key: Option<String>,
+    bucket: Option<String>,
+    auth_bucket: Option<String>,
+}
+
+impl Debug for Builder {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Builder")
+            .field("root", &self.root)
+            .field("endpoint", &self.endpoint)
+            .field("access_key", &"<redacted>")
+            .field("secret_key", &"<redacted>")
+            .field("bucket", &self.bucket)
+            .field("auth_bucket", &self.auth_bucket)
+            .finish()
+    }
+}
+
+impl Builder {
+    /// Set root of this backend.
+    ///
+    /// All operations will happen under this root.
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        if !root.is_empty() {
+            self.root = Some(root.to_string())
+        }
+
+        self
+    }
+
+    /// Set endpoint of this backend.
+    ///
+    /// Endpoint must be full uri, e.g.
+    ///
+    /// - `https://bucketname.obs.cn-north-4.myhuaweicloud.com`
+    pub fn endpoint(&mut self, endpoint: &str) -> &mut Self {
+        if !endpoint.is_empty() {
+            self.endpoint = Some(endpoint.trim_end_matches('/').to_string());
+        }
+
+        self
+    }
+
+    /// Set access_key of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn access_key(&mut self, access_key: &str) -> &mut Self {
+        if !access_key.is_empty() {
+            self.access_key = Some(access_key.to_string());
+        }
+
+        self
+    }
+
+    /// Set secret_key of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn secret_key(&mut self, secret_key: &str) -> &mut Self {
+        if !secret_key.is_empty() {
+            self.secret_key = Some(secret_key.to_string());
+        }
+
+        self
+    }
+
+    /// Set bucket of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn bucket(&mut self, bucket: &str) -> &mut Self {
+        if !bucket.is_empty() {
+            self.bucket = Some(bucket.to_string());
+        }
+
+        self
+    }
+
+    /// Set bucket in CanonicalizedResource of this backend.
+    /// - If it is set, we will take user's input first.
+    /// - If not, we will try to load it from environment.
+    pub fn auth_bucket(&mut self, auth_bucket: &str) -> &mut Self {
```

Comment:
We don't need `auth_bucket`.

- If user doesn't input endpoint, use bucket for `signer`
- If user inputs endpoint, use endpoint's domain for signer.

Reply:
[obs docs: Authentication of Signature in a Header](https://support.huaweicloud.com/intl/en-us/api-obs/obs_04_0010.html) specifies how to calculate CanonicalizedResource.

```
Indicates the OBS resource specified by an HTTP request. This parameter is constructed as follows:

<Bucket name + Object name> + [Subresource 1] + [Subresource 2] + ...

Bucket name and object name
If the bucket is bound to a user domain name, use the user domain name as the bucket name, for example, /obs.ccc.com/object. obs.ccc.com is the user domain name bound to the bucket. If no object name is specified, the entire bucket is listed, for example, /obs.ccc.com/.
If you do not access OBS using a user domain name, this field is in the format of /bucket/object. If no object name is specified, for example, /bucket/, the entire bucket is listed. If bucket name is not specified either, the value of this field is /.

...
```

For user domain, we use the domain as `bucket` in Signer, and `endpoint` in Backend, like:

```
url: https://obs.cccccc.com/object.txt
             |- endpoint -|
CanonicalizedResource: /obs.cccccc.com/object.txt
                        |- endpoint -|
(use user domain as bucket name in CanonicalizedResource)
```

For default domain, we use the bucket as `bucket` in Signer, and `bucket` plus `endpoint` in Backend. like:

```
url: https://bucket_namee.obs.cn-north-4.myhuaweicloud.com/object.txt
             |- bucket -| |------- endpoint -------------|
CanonicalizedResource: /bucket_namee/object.txt
                        |- bucket -|
```

We don't know which mode of endpoint is using, so we have to ask the user to set `bucket` param for url concatentaion and `auth_bucket` for CanonicalizedResource.

Reply:
Users must input **valid** endpoints as we required. We don't need to accept all endpoint string.

As for now, we only accept two kinds endpoints:

- huaweicloud endpoint: `https://obs.<region>.myhuaweicloud.com`
- users own endpoint: `https://obs.cccccc.com`

For huaweicloud endpoint, we will append bucket for users. For user's own endpoint, we will use the domain directly.

Reply:
Fixed, PTAL!